src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -133331,6 +133331,160 @@ internal interface IB : IA { }
+        [WorkItem(46342, "https://github.com/dotnet/roslyn/issues/46342")]
+        public void LambdaNullReturn_02()
+        {
+            var source =
+@"#nullable enable
+using System;
+class Program
+{
+    static void F<T>(Func<T> f)
+    {
+    }
+    static void M<T>(bool b, T t) where T : class?
+    {
+        F(() =>
+        {
+            if (b) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(46342, "https://github.com/dotnet/roslyn/issues/46342")]
+        public void LambdaNullReturn_03()
+        {
+            var source =
+@"#nullable enable
+using System;
+class Program
+{
+    static void F<T>(Func<T> f)
+    {
+    }
+    static void M<T>(T? t) where T : class
+    {
+        F(() =>
+        {
+            if (t is null) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(46342, "https://github.com/dotnet/roslyn/issues/46342")]
+        public void LambdaNullReturn_04()
+        {
+            var source =
+@"#nullable enable
+using System;
+class Program
+{
+    static void F<T>(Func<T> f)
+    {
+    }
+    static void M<T>(T? t) where T : class?
+    {
+        F(() =>
+        {
+            if (t is null) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(46342, "https://github.com/dotnet/roslyn/issues/46342")]
+        public void LambdaNullReturn_05()
+        {
+            var source =
+@"#nullable enable
+using System;
+using System.Threading.Tasks;
+class Program
+{
+    static void F<T>(Func<Task<T>> f)
+    {
+    }
+    static void M<T>(T? t) where T : class
+    {
+        F(async () =>
+        {
+            await Task.Yield();
+            if (t is null) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(46342, "https://github.com/dotnet/roslyn/issues/46342")]
+        public void LambdaNullReturn_06()
+        {
+            var source =
+@"#nullable enable
+using System;
+using System.Threading.Tasks;
+class Program
+{
+    static void F<T>(Func<Task<T>> f)
+    {
+    }
+    static void M<T>(T? t) where T : class?
+    {
+        F(async () =>
+        {
+            await Task.Yield();
+            if (t is null) return null;
+            return t;
+        });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyEmitDiagnostics();
+        }
+
```

Comment:
Could we add similar tests with `default`? File an issue if misbehaving.

Reply:
`return default;` results in "warning CS8603: Possible null reference return" with 16.7 and this branch.
`return default(T);` compiles without warnings with both. 

I'll add tests for those cases in a separate PR.